### PR TITLE
[AArch64,ELF] Allow implicit $d/$x at section beginning

### PR DESCRIPTION
**Pull request by MaskRay**
The start state of a new section is `EMS_None`, often leading to a
$d/$x at offset 0. Introduce a MCTargetOption/cl::opt
"implicit-mapsyms" to allow an alternative behavior
(https://github.com/ARM-software/abi-aa/issues/274):

* Set the start state to `EMS_Data` or `EMS_A64`.
* For text sections, add an ending $x only if the final data is not instructions.
* For non-text sections, add an ending $d only if the final data is not data commands.

```
.section .text.1,"ax"
nop
// emit $d
.long 42
// emit $x

.section .text.2,"ax"
nop
```

This new behavior decreases the .symtab size significantly:

```
% bloaty a64-2/bin/clang -- a64-0/bin/clang
    FILE SIZE        VM SIZE
 --------------  --------------
  -5.4% -1.13Mi  [ = ]       0    .strtab
 -50.9% -4.09Mi  [ = ]       0    .symtab
  -4.0% -5.22Mi  [ = ]       0    TOTAL
```

---

This scheme works as long as the user can rule out some error scenarios:

* .text.1 assembled using the traditional behavior is combined with .text.2 using the new behavior
* A linker script combining non-text sections and text sections. The
  lack of mapping symbols in the non-text sections could make them
  treated as code, unless the linker inserts extra mapping symbols.

The above mix-and-match scenarios aren't an issue at all for a
significant portion of users.

A text section may start with data commands in rare cases (e.g.
-fsanitize=function) that many users don't care about. When combing
`(.text.0; .word 0)` and `(.text.1; .word 0)`, the ending $x of .text.0
and the initial $d of .text.1 may have the same address. If both
sections reside in the same file, ensure the ending symbol comes before
the initial $d of .text.1, so that a dumb linker respecting the symbol
order will place the ending $x before the initial $d.

Disassemblers using stable sort will see both symbols at the same
address, and the second will win.

When section ordering mechanisms (e.g. --symbol-ordering-file,
--call-graph-profile-sort, `.text : { second.o(.text) first.o(.text) }`)
are involved, the initial data in a text section following a text
section with trailing data could be misidentified as code, but the issue
is local and the risk could be acceptable.

### CODE REVIEW

**Comment by MaskRay** [comment]
The new option "optimize-mapping-symbols" is for llc/llvm-mc/etc. `clang -mllvm -optimize-mapping-symbols` does not get the option. Another patch will be needed to enable a `-Wa,` option, similar to #97378 (CREL).

It's desirable to coordinate with the GNU binutils project on the naming of this option.

It's worth noting that the `EMS_A64` start state modification for text sections currently deviates from the standard (https://github.com/ARM-software/abi-aa/issues/274), but could become compliant when the ABI is updated.

**Comment by MaskRay** [comment]
As suggested by a binutils global maintainer, Jan Beulich, `optimize-mapping-symbols` has been renamed to `optimize-mapsyms`.
I will add clangDriver support for `-Wa,--optimize-mapsyms` once this is approved.

While binutils aarch64 maintainer Richard hasn't yet endorsed the solution, I believe there's significant value in having this option
for users who don't (a) mix data and code or (b) mix data-in-code text sections before or after this patch.

This feature is for LLVM 20, not for 19.x release, so that I will have more time to convince Richard.
(Although there is a desire to have this early to benefit my downstream...)

**Comment by MaskRay** [comment]
🥳🥳

**Comment by MaskRay** [comment]
Ping:)

**Comment by smithp35** [comment]
Apologies for the delay in responding.

I'm uncomfortable with symbols defined at the end of the section as these fall through into the following section. I have seen that cause trouble before with 0 sized sections as contradictory mapping symbols at the same address are not well defined by the ABI. While it is true that the most likely implementation of a linker would order mapping symbols such that a trailing mapping symbol would get sorted before a mapping symbol at offset 0 in the next section; I'm not sure we can guarantee it, for example a non-stable sort of the symbols by address could break that. I also know of at least one linker that tries to optimize mapping symbols by removing redundant mapping symbols that will give an error message if it spots two contradictory mapping symbols at the same address.

I think it would be better to just omit the trailing $x mapping symbol as I expect it would cause more problems than it will solve. As `-foptimize-mapsyms` is opt-in and not the default, a user could be responsible for dealing with any corner-cases that arise with $d as the last mapping symbol in code-sections. For clang it looks like this is very rare. It is more common in GCC as it uses literal pools for -mcmodel=large. Clang uses 4 mov instructions instead. For what it is worth a colleague of mine is looking into adding execute-only support in AArch64, this largely means leaving code-generation alone, but giving an error if data is used in an executable section, this would guarantee no corner cases.

Going back to the ABI perhaps it would be worth looking at symbols rather than sections. Essentially can we make a symbol of type STT_FUNC with bit 0 clear (thinking of Morello, which uses 1 for C64 instructions) an implicit $x and a symbol of type STT_OBJECT an implicit $d, with all other symbols being neutral? This would solve the trailing $d problem for compiled code and well written assembly code (people are lazy about setting the type of symbols).

**Comment by MaskRay** [comment]
Thank you for your feedback. You suggest a scheme where the trailing $x mapping symbol is omitted to prevent potential contradictory mapping symbols at the same address.

```
.section .text.1,"ax"
nop
// emit $d
.long 42
// do not emit $x
```

This requires a smart linker that inserts extra mapping symbols.
While I understand the benefits of this approach, it's not suitable for all users.
Those who can tolerate potential mapping symbol conflicts or have tooling to handle them (even if not well defined) might prefer an ending mapping symbol.

I believe it's very difficult for Clang to emit trailing data in a text section.
However, hand-written assembly might do so, and in the absence of a smart linker, large regions of .text could be misidentified as data.

While I'm unfamiliar with Morello, omitting a starting label (often the function name) is possible in hand-written assembly. This disqualifies approaches that rely on start label st_vaue.
Additionally, I'm less comfortable with ABIs that use st_value&1 for state information (like Thumb and MicroMips).
This can lead to ambiguity when defining label differences (A-B vs. A-B+C) in assembly code. https://reviews.llvm.org/D157655

---

**Is it feasible to have a cl::opt (not -Wa,--foo) to have the ending mapping symbol behavior?**

I understand the preference for a limited number of schemes, but I anticipate that we may end up with three distinct options as users have different needs:
the current approach (there are already differences with gas; LLVM AArch32 does this https://reviews.llvm.org/D30724), the proposed patch, and a variation of the patch without the ending mapping symbol.

**Comment by smithp35** [comment]
Morello definitely couldn't use this option https://github.com/ARM-software/abi-aa/blob/main/aaelf64-morello/aaelf64-morello.rst#mapping-symbols . Having said that Morello is not part of upstream LLVM and is an architecture experiment.

Yes I don't think there will be a single option that is going to work for everyone.
| option | impact |
|--------|---------|
| default | covers all cases but too many mapping symbols |
| omit leading $x, trailing $x if required | could cause $x and $d at same address, error in some tools. or be incorrect when a data section follows a code section with a trailing $x |
| omit leading $x, no trailing $x | incorrect when data is last in the section and code follows in next section. Will have double $x when combined with the default option. |
| imply $x from STT_FUNC, $d from STT_OBJECT | no guarantee that symbol is first, and user has typed symbol properly. |

On reflection I think putting the trailing $x in is probably the least worst optimised solution as I expect it is the least likely to cause problems.

Is there a strong objection to optimising this at link time? There would be some run-time overhead to scan for redundant $x and removing them, but this would always give the same answer as the current default.

**Comment by smithp35** [comment]
Some thoughts after thinking about this on the weekend and talking to some colleagues.

Apologies for the length of the response, a brief summary:
* I think this could be a useful option if projects build all their objects with it enabled. Would be good to know from potential users that they would enable this option in this review? So far I think it is only me that has commented.
* My main concern is binary distribution of objects with it enabled, and what future choices this might rule out.
* I'm still nervous about trailing mapping symbols, particularly with `-fsanitize=undefined` or `-fsanitize=kcfi`.
* Can we alleviate these concerns by marking object files that use this mapping symbol convention. That way we could fix them up later if needed. 
* Is there an alternative way to do link-time culling of mapping symbols that runs faster?

In more detail.

The pragmatist in me says that this is a useful (possibly llvm only [1]) option for people with large builds to enable an alternative mapping symbol definition than the ABI and accept the consequences of that as well as the benefits. The idealist in me says that if this propagates out into binary distributions across multiple projects then we may have incompatibilities with GNU and while there are no plans to add one, it may rule out something like a second instruction set (like Morello or Thumb [2]).
[1] the comments on the binutils mailing list don't look enthusiastic.
[2] We can't know which instruction set is implicit.

A colleague working on Bolt mentioned that there could be a case for more mapping symbols to help binary analysis, this implies that there could be more mapping symbol density options than just optimize and default. He also mentioned that recording the mapping symbol level choice in the ELF file (mechanism TBD) might be a way around the future compatibility problem, and could extend to the additional mapping symbol cases. Does recording the use of the alternative mapping symbol mechanism appeal at all?

I'm still nervous about adding trailing `$x`. I had thought that AArch64 executable sections with a leading `$d` would be very rare, but I had forgotten about `-fsanitize=kcfi` and `-fsanitize=undefined`. These all lead with a `$d`. This makes collisions between trailing `$x` and `$d` more likely than I'd first thought.

I also don't think we can rely on the dumb linker collation order putting the trailing `$x` before the leading `$d` of the following section. For example, assuming we have `lld first.o second.o` with the following linker script fragment: 
```
  .text : { second.o(.text) first.o(.text) }
```
If `.text` in first.o leads with a `$d`; like when `-fsanitize=kcfi` is used, and `.text` in second.o has a trailing `$x` then I think when we copy local symbols into the symbol table first.o will get copied first, which will be maintained by stable_partition so we'll get `$d` `$x` at the same address with the `$x` . This is all theoretical, I haven't tried testing it yet.

As an aside, I'm surprised the linker removal of redundant mapping symbols took so long. It looks like we already stable_partition the static symbol table in `void SymbolTableBaseSection::finalizeContents()`. Did you modify that to stable_sort rather than partition, or just stable_sort the local symbols after the partition?

Another possibility that I think would work in the majority of AArch64 cases without costing too much extra link time, which I think could be done before the stable_partition (to speed that up) is to see if each OutputSection always has the same mapping symbol type. For example if the `.text` section only has `$x` mapping symbols, which would be the most common case, then we can replace these with a single `$x` mapping symbol. If there any `$d` in there bail out. I think this could be done much more cheaply than 5% link time.

**Comment by statham-arm** [comment]
> Can we alleviate these concerns by marking object files that use this mapping symbol convention. That way we could fix them up later if needed.

That was my first thought, too.

As far as I can see, this option is essentially inventing a nonstandard variation on the Arm ELF spec, which is not fully compatible with the standard one, as the PR description mentions:

> This scheme works as long as the user can rule out some error scenarios:
> 
> * .text.1 assembled using the traditional behavior is combined with .text.2 using the new behavior

Currently, as far as I can see, the only way that a user _can_ rule out this error scenario is to use their own memory and self-discipline to keep track of which objects (or which libraries, or which whole projects) they're building using this nonstandard flag.

Surely it would be better to mark object files to indicate explicitly that they follow this nonstandard ABI. Then at the very least a linker can detect a mixture of the two and complain, even if it can't automatically compensate. That way the user doesn't have to do it all by self-discipline.

**Comment by MaskRay** [comment]
> Some thoughts after thinking about this on the weekend and talking to some colleagues.
> 
> Apologies for the length of the response, a brief summary:
> 
> * I think this could be a useful option if projects build all their objects with it enabled. Would be good to know from potential users that they would enable this option in this review? So far I think it is only me that has commented.
> * My main concern is binary distribution of objects with it enabled, and what future choices this might rule out.
> * I'm still nervous about trailing mapping symbols, particularly with `-fsanitize=undefined` or `-fsanitize=kcfi`.
> * Can we alleviate these concerns by marking object files that use this mapping symbol convention. That way we could fix them up later if needed.
> * Is there an alternative way to do link-time culling of mapping symbols that runs faster?
> 
> In more detail.
> 
> The pragmatist in me says that this is a useful (possibly llvm only [1]) option for people with large builds to enable an alternative mapping symbol definition than the ABI and accept the consequences of that as well as the benefits. The idealist in me says that if this propagates out into binary distributions across multiple projects then we may have incompatibilities with GNU and while there are no plans to add one, it may rule out something like a second instruction set (like Morello or Thumb [2]). [1] the comments on the binutils mailing list don't look enthusiastic. [2] We can't know which instruction set is implicit.

Thanks for the comprehensive summary. I concur with your analysis, including the subsequent points. I believe some users, like mine, are willing to accept the consequences.

> A colleague working on Bolt mentioned that there could be a case for more mapping symbols to help binary analysis, this implies that there could be more mapping symbol density options than just optimize and default. He also mentioned that recording the mapping symbol level choice in the ELF file (mechanism TBD) might be a way around the future compatibility problem, and could extend to the additional mapping symbol cases. Does recording the use of the alternative mapping symbol mechanism appeal at all?
> 
> I'm still nervous about adding trailing `$x`. I had thought that AArch64 executable sections with a leading `$d` would be very rare, but I had forgotten about `-fsanitize=kcfi` and `-fsanitize=undefined`. These all lead with a `$d`. This makes collisions between trailing `$x` and `$d` more likely than I'd first thought.

Yes. [-fsanitize=function](https://maskray.me/blog/2022-12-18-control-flow-integrity#fsanitizefunction) and `-fsanitize=kcfi` emit executable sections with a leading `$d`.

> I also don't think we can rely on the dumb linker collation order putting the trailing `$x` before the leading `$d` of the following section. For example, assuming we have `lld first.o second.o` with the following linker script fragment:
> 
> ```
>   .text : { second.o(.text) first.o(.text) }
> ```
> 
> If `.text` in first.o leads with a `$d`; like when `-fsanitize=kcfi` is used, and `.text` in second.o has a trailing `$x` then I think when we copy local symbols into the symbol table first.o will get copied first, which will be maintained by stable_partition so we'll get `$d` `$x` at the same address with the `$x` . This is all theoretical, I haven't tried testing it yet.

Agreed, this is a potential downside similar to --symbol-ordering-file and [--call-graph-profile-sort](https://maskray.me/blog/2020-11-15-explain-gnu-linker-options#call-graph-profile-sort).
lld will place first.o symbols before second.o symbols, potentially causing the undesired `$d $x` order.
I know my user is willing to accept the consequence.

Fortunately, this risk seems manageable for some users. First, text sections with trailing data are extremely rare.
Even if such a section is followed by a `-fsanitize=function` section (`.long 0xc105cafe; .long 2772461324; code`),
and the two words are misidentified as code, the following code won't, making this an acceptable risk for some users.

> As an aside, I'm surprised the linker removal of redundant mapping symbols took so long. It looks like we already stable_partition the static symbol table in `void SymbolTableBaseSection::finalizeContents()`. Did you modify that to stable_sort rather than partition, or just stable_sort the local symbols after the partition?

I have developed this hack https://github.com/MaskRay/llvm-project/tree/lld-mapsym , 5% slowdown.

```patch
diff --git a/lld/ELF/SyntheticSections.cpp b/lld/ELF/SyntheticSections.cpp
index c27ab2b67dc2..d8e4bc976016 100644
--- a/lld/ELF/SyntheticSections.cpp
+++ b/lld/ELF/SyntheticSections.cpp
@@ -2172,4 +2172,9 @@ void SymbolTableBaseSection::sortSymTabSymbols() {
   getParent()->info = numLocals + 1;

+  for (auto &s : symbols)
+    s.va = s.sym->getVA();
+  std::stable_sort(symbols.begin(), e,
+                   [](auto &l, auto &r) { return l.va < r.va; });
+
   // We want to group the local symbols by file. For that we rebuild the local
   // part of the symbols vector. We do not need to care about the STT_FILE
@@ -2190,5 +2195,5 @@ void SymbolTableBaseSection::addSymbol(Symbol *b) {
   // Adding a local symbol to a .dynsym is a bug.
   assert(this->type != SHT_DYNSYM || !b->isLocal());
-  symbols.push_back({b, strTabSec.addString(b->getName(), false)});
+  symbols.push_back({b, strTabSec.addString(b->getName(), false), 0});
 }

@@ -2266,5 +2271,5 @@ template <class ELFT> void SymbolTableSection<ELFT>::writeTo(uint8_t *buf) {
       if (isDefinedHere) {
         eSym->st_shndx = shndx;
-        eSym->st_value = sym->getVA();
+        eSym->st_value = ent.va ? ent.va : sym->getVA();
         // Copy symbol size if it is a defined symbol. st_size is not
         // significant for undefined symbols, so whether copying it or not is up
@@ -2492,5 +2497,5 @@ void GnuHashTableSection::addSymbols(SmallVectorImpl<SymbolTableEntry> &v) {
   v.erase(mid, v.end());
   for (const Entry &ent : symbols)
-    v.push_back({ent.sym, ent.strTabOffset});
+    v.push_back({ent.sym, ent.strTabOffset, 0});
 }

diff --git a/lld/ELF/SyntheticSections.h b/lld/ELF/SyntheticSections.h
index d4169e1e1aca..c7e385119c03 100644
--- a/lld/ELF/SyntheticSections.h
+++ b/lld/ELF/SyntheticSections.h
@@ -642,4 +642,5 @@ struct SymbolTableEntry {
   Symbol *sym;
   size_t strTabOffset;
+  uint64_t va;
 };
```

> Another possibility that I think would work in the majority of AArch64 cases without costing too much extra link time, which I think could be done before the stable_partition (to speed that up) is to see if each OutputSection always has the same mapping symbol type. For example if the `.text` section only has `$x` mapping symbols, which would be the most common case, then we can replace these with a single `$x` mapping symbol. If there any `$d` in there bail out. I think this could be done much more cheaply than 5% link time.

This approach (deleting extra mapping symbols) is for the default case. Bailing out could hinder a significant size optimization for executables with minor data-in-code elements.
```
default	| covers all cases but too many mapping symbols
```

Alternatively, removing leading `$x` can reduce relocatable file size, which will be preferred by my users.
We have two choices:
```
omit leading $x, trailing $x if required	| could cause $x and $d at same address, error in some tools. or be incorrect when a data section follows a code section with a trailing $x
omit leading $x, no trailing $x	| incorrect when data is last in the section and code follows in next section. Will have double $x when combined with the default option.
```

The first choice has been discussed above. Technically a linker could delete `$x` in a text section if it shares the address with a `$d`.
This feature could be implemented with minimal code, but I believe my users won't require this level of precision.

The second choice "omit leading $x, no trailing $x" requires a smart linker that builds a output-section-to-symbol map (`.text => $x $d $x $d $x`) and inspects input section boundaries.
I believe there is significant performance overhead (sorting), increased memory consumption (tracking input section boundaries without stealing one bit from `lld::ELF::Symbol`), and potential code complexity.

**Comment by MaskRay** [comment]
> > Can we alleviate these concerns by marking object files that use this mapping symbol convention. That way we could fix them up later if needed.
> 
> That was my first thought, too.
> 
> As far as I can see, this option is essentially inventing a nonstandard variation on the Arm ELF spec, which is not fully compatible with the standard one, as the PR description mentions:
> 
> > This scheme works as long as the user can rule out some error scenarios:
> > 
> > * .text.1 assembled using the traditional behavior is combined with .text.2 using the new behavior
> 
> Currently, as far as I can see, the only way that a user _can_ rule out this error scenario is to use their own memory and self-discipline to keep track of which objects (or which libraries, or which whole projects) they're building using this nonstandard flag.

You're absolutely right. The nonstandard option is invaluable for those with full control over their toolchain, no reliance on weird relocatable files, and a strong focus on minimizing both relocatable and executable sizes.
This aligns with the needs of (a) large-scale build environments, (b) huge open-source projects, and (c) distributions like Gentoo.
(I haven't used Gentoo for a while, but I would certainly use this option to make my builds faster and smaller.)

> Surely it would be better to mark object files to indicate explicitly that they follow this nonstandard ABI. Then at the very least a linker can detect a mixture of the two and complain, even if it can't automatically compensate. That way the user doesn't have to do it all by self-discipline.

While marking object files could be beneficial, it's essential to avoid overly restrictive compatibility checks. False positives could hinder development.
(I actually wanted to write a blog post about compatibility check and inconvenience, but I kept postponing the writing...)

For example, the following code assembled with a traditional assembler could get marked to be incompatible with `.text; .nop` using the nonstandard behavior.
```
.text
nop
.long 0
```

But pure code should not trigger incompatiblity warning.
```
.text
nop
```

**Comment by smithp35** [comment]
I don't think we need to use an object marking to do a compatibility check, while that is possible I don't think it is necessary in the majority of the cases.

What it does do is give us the chance to identify objects with the alternative marking scheme and unambiguously know that an executable section with no mapping symbols is either data or AArch64 code. This could help tools if the difference became important.

While I don't want to speculate too much over how to mark an object, I'm wondering if it is something you're willing to entertain?

As an aside; Arm's original ELF spec for its proprietary toolchain prior to the ABI in 2004 used to define a couple of ELF header flags:
* `EF_ARM_MAPSYMSFIRST` all mapping symbols form a subsection which is first in the static symbol table, all remaining symbols form a subsection. The first symbol in the symbol table is `$m` the value of which was the number of mapping symbols, permitting a consumer to skip to the first non-mapping symbol.
* `EF_ARM_SYMSARESORTED` symbols in a subsection are sorted in ascending address order.

These were rejected when the AArch32 ABI was published as specific to Arm's proprietary toolchain.

**Comment by smithp35** [comment]
To give you a summary of my thoughts so far:
* I don't think we'll be able to change the AArch64 ABI as I don't think there's enough consensus on the GNU side for a relocatable object based solution. 
* An alternative, likely LLVM only, opt-in option for users with large projects, in control of their own toolchain to reduce mapping symbols could be valuable to those projects.
* I'd like to discourage use of the opt-in option where portability of the relocatable objects is a concern. For example, while there are no plans to do so, I couldn't guarantee the AArch64 ABI might make changes that would retain backwards compatibility.
* If the option becomes LLVM only then I'd prefer we gave the option a more neutral name. For example `-mimplicit-mapsyms` or `-malternative-mapsyms` as `-moptimize-mapsyms` doesn't imply a, small, trade-off in portability.
* I'd like to see the help for the option mention compatibility. For example "Allow mapping symbol at section beginning to be implicit, lowers number of mapping symbols at the expense of some portability. Recommended for large projects that can build all their objects using this option." 
* I'd like to find a way of recording that implicit mapping symbols have been used, I don't think we need a compatibility check.

**Comment by MaskRay** [comment]
Renamed to "implicit-mapsyms"

---

> To give you a summary of my thoughts so far:
> 
> * I don't think we'll be able to change the AArch64 ABI as I don't think there's enough consensus on the GNU side for a relocatable object based solution.

Ack.

> * An alternative, likely LLVM only, opt-in option for users with large projects, in control of their own toolchain to reduce mapping symbols could be valuable to those projects.

Looks good. 

> * I'd like to discourage use of the opt-in option where portability of the relocatable objects is a concern. For example, while there are no plans to do so, I couldn't guarantee the AArch64 ABI might make changes that would retain backwards compatibility.

I agree with your point about discouraging its use where portability is crucial.

Does the last sentence read as: the AArch64 ABI might make changes to break backwards compatibility. But with what?

As is, the proposed `implicit-mapsyms` option is non-conforming per
"A section that contains instructions must have a mapping symbol defined at the beginning of the section. If a section cont     ains only data no mapping symbol is required.

I agree that users of the opt-in option should well understand the risks.

> * If the option becomes LLVM only then I'd prefer we gave the option a more neutral name. For example `-mimplicit-mapsyms` or `-malternative-mapsyms` as `-moptimize-mapsyms` doesn't imply a, small, trade-off in portability.

Looks good.
The MCTargetOption/cl::opt option (for llc/llvm-mc) has been renamed to `implicit-mapsyms`.
Another patch will be needed to add a driver option, perhaps `-Wa,-mimplicit-mapsyms`.

> * I'd like to see the help for the option mention compatibility. For example "Allow mapping symbol at section beginning to be implicit, lowers number of mapping symbols at the expense of some portability. Recommended for large projects that can build all their objects using this option."

Thanks for the suggestion. Improved the message at `llvm/lib/MC/MCTargetOptionsCommandFlags.cpp`.

> * I'd like to find a way of recording that implicit mapping symbols have been used, I don't think we need a compatibility check.

Looks good. Recording the mapping symbol scheme might depend on `SHT_AARCH64_ATTRIBUTES`, which has no GNU/LLVM toolchain implementation yet.

> https://github.com/llvm/llvm-project/pull/99718#issuecomment-2286565171 While I don't want to speculate too much over how to mark an object, I'm wondering if it is something you're willing to entertain?

Yes, I'm open to exploring that once the base mechanism is available.

**Comment by smithp35** [comment]
Just a quick comment to say thanks for the updates. I ran out of time to look today, will try and take a look tomorrow.

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp
Comment (smithp35) [inline llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:338]: It looks like we have a count of symbols for every section. For a large object with lots of sections (coming from say LTO with -ffunction-sections) then if we expect trailing symbols to be rare, then it could be better to test if the section has a trailing mapping symbol early, and only retain counts for the  sections that we are adding to.

May not be worth it for the extra complexity though.

File: llvm/test/MC/AArch64/mapping-across-sections.s
Comment (smithp35) [inline llvm/test/MC/AArch64/mapping-across-sections.s:2]: Not a comment on this test. I think it would be worth a test, probably in lld, or possibly via yaml2obj to test what the disassembler does when it gets two mapping symbols at the same address.

**Review by smithp35 — COMMENTED** [review]
These are comments on the implementation/test only.

**Comment by smithp35** [comment]
> Renamed to "implicit-mapsyms"
> 
> > To give you a summary of my thoughts so far:
> > 
> > * I don't think we'll be able to change the AArch64 ABI as I don't think there's enough consensus on the GNU side for a relocatable object based solution.
> 
> Ack.
> 
> > * An alternative, likely LLVM only, opt-in option for users with large projects, in control of their own toolchain to reduce mapping symbols could be valuable to those projects.
> 
> Looks good.
> 
> > * I'd like to discourage use of the opt-in option where portability of the relocatable objects is a concern. For example, while there are no plans to do so, I couldn't guarantee the AArch64 ABI might make changes that would retain backwards compatibility.
> 
> I agree with your point about discouraging its use where portability is crucial.
> 
> Does the last sentence read as: the AArch64 ABI might make changes to break backwards compatibility. But with what?
> 
If another ISA like Thumb is to Arm is added to AArch64 [*] then we wouldn't be able to tell which instruction set to start with for an executable section. There are likely some implicit assumptions we can make if we're sensible, such as defaulting to AArch64 if there's no mapping symbol, and if the other ISA is used it needs a mapping symbol. I suspect that llvm tools could be made to work, however it could get messy with GNU and other tools. 

[*] Morello's hybrid mode does this. A subset of the CHERI instructions can be used in EM_AARCH64 after some transitioning like Arm to Thumb. The majority of experimentation with Morello has been in pure capability mode which is a separate machine from AARCH64.

> Looks good. Recording the mapping symbol scheme might depend on `SHT_AARCH64_ATTRIBUTES`, which has no GNU/LLVM toolchain implementation yet.
> 
> > [#99718 (comment)](https://github.com/llvm/llvm-project/pull/99718#issuecomment-2286565171) While I don't want to speculate too much over how to mark an object, I'm wondering if it is something you're willing to entertain?
> 
> Yes, I'm open to exploring that once the base mechanism is available.
> 

I'm expecting a build attributes implementation in GNU fairly shortly for GCS assuming it can hit early enough for the GCC release cycle. It may have to slip to the next GCC release if it doesn't make it in time though. I'm hoping that we can get a LLVM implementation started for LLVM 20.

File: llvm/test/MC/AArch64/mapping-across-sections.s
Comment (MaskRay) [reply llvm/test/MC/AArch64/mapping-across-sections.s:2]: Added lld/test/ELF/aarch64-mapsyms-implicit.s

**Comment by smithp35** [comment]
Apologies for the delay in responding, I'm aiming to take a look tomorrow, as well as the other ones in the series.

**Review by smithp35 — COMMENTED** [review]
Thanks for the updates.

Although possibly not for this patch. It looks llvm-objdump (and gnu objdump) look to reorder $x after $d when the symbols are at the same address. This looks like worse case behaviour.

I think we could address this by looking at llvm-objdump, or we could have lld filter out the clashes.

The other tool that I thought could be affected by this is Bolt. It uses mapping symbols to determine the location of constant islands. It looks to do this with CodeOffsets and DataOffsets (some references below). Are your large programs likely to be post-processed by Bolt? 

Some references:
https://github.com/llvm/llvm-project/blob/main/bolt/include/bolt/Core/BinaryFunction.h#L1972
https://github.com/llvm/llvm-project/blob/main/bolt/lib/Core/BinaryEmitter.cpp#L539
https://github.com/llvm/llvm-project/blob/main/bolt/lib/Core/BinaryFunction.cpp#L1049

Apologies to keep bringing these up. What I want to avoid is people finding other llvm tools subtly broken. I don't think that these are a blocker for implicit mapsyms, but it should influence the level of warning we give for the option. We may want to call out Bolt specifically as a tool that could be affected for example.

I think the code here is good for llvm-mc. I don't have any more comments on this patch.

I've got a last couple of questions:
* Do you want to update llvm-objdump as whatever collation order the linker uses, it is going to order `$d` before `$x`? I think that could be done in different patch.
* If Bolt could get confused by two symbols at the same address I think we'll need to call that out in the clang driver help text. Do you think it would be worth an lld patch to remove (or make both symbols the same) when there is a clash? That would make Bolt work for a LLVM based toolchain.

**Comment by smithp35** [comment]
Just a quick note that I'll be out of office from Friday 23rd. Will be back on Tuesday 27th so I may be a bit slow to respond.

**Comment by MaskRay** [comment]
> Although possibly not for this patch. It looks llvm-objdump (and gnu objdump) look to reorder $x after $d when the symbols are at the same address. This looks like worse case behaviour.

Yes, the -fsanitize=function and -fsanitize=kcfi cases that many people don't use.
In particular, as a contributor of -fsanitize=kcfi, I think the feature should not be used for userspace programs.
They are specific to the Linux kernel, which expects a specific conditional branch instruction and a specific trap instruction.
(FreeBSD kernel could adopt it if they decide to do something similar to the Linux kernel.)

> I think we could address this by looking at llvm-objdump, or we could have lld filter out the clashes.

Yes. The test intends to demonstrate the weakness.

> The other tool that I thought could be affected by this is Bolt. It uses mapping symbols to determine the location of constant islands. It looks to do this with CodeOffsets and DataOffsets (some references below). Are your large programs likely to be post-processed by Bolt?

No, not post-processed by BOLT.

> Do you want to update llvm-objdump as whatever collation order the linker uses, it is going to order $d before $x? I think that could be done in different patch.

Yes, I will take a stab.

* If Bolt could get confused by two symbols at the same address I think we'll need to call that out in the clang driver help text. Do you think it would be worth an lld patch to remove (or make both symbols the same) when there is a clash? That would make Bolt work for a LLVM based toolchain.

I want to change `clang -Wa,` options to use TableGen so that we can place additional help messages.
Right now `-Wa,` options don't have a way to provide messages.
`clang/docs/UsersManual.rst` might also be a choice, if we don't add messages to `--help`.

Currently, we could only rely on the `cl::opt` message. I feel that BOLT is a very specific detail that cannot be well described a brief message...

---

Enjoy your vacation! I'll take a flight on Friday and will be slow to respond for about 3 weeks.

**Review by smithp35 — APPROVED** [review]
Thanks for all the feedback. I don't have any more feedback on this specific patch.

Hopefully we'll be able to improve llvm-objdump and possibly lld to avoid the corner cases.

Thanks for letting us know you'll be slow to respond over the next few weeks.

**Pull request merged**
